torch/csrc/dynamo/compiled_autograd_py.cpp

```diff
@@ -0,0 +1,526 @@
+  THPObjectPtr pyresult(check(PyObject_CallMethodObjArgs(
+      self, method_name, pyinput.get(), pysizeinput.get(), NULL)));
+
+  PyObject *fake_inputs, *fake_sizes;
+  check(PyArg_ParseTuple(pyresult.get(), "OO", &fake_inputs, &fake_sizes));
+  return TraceState(
+      unwrap_variable_list(fake_inputs),
+      cache.unwrap_dynamic_inputs(fake_sizes),
+      compiler_call.accumulate_grad,
+      num_outputs);
+}
+
+static PyObject* call_end_capture(PyObject* self, const variable_list& inputs) {
+  static PyObject* method_name = PyUnicode_InternFromString("end_capture");
+  THPObjectPtr pyinput(wrap_variable_list(inputs));
+  return check(PyObject_CallMethodOneArg(self, method_name, pyinput.get()));
+}
+
+variable_list compiled_autograd(
+    const std::shared_ptr<Node>& graph_root,
+    GraphTask& graph_task,
+    bool accumulate_grad,
+    const edge_list& output_edges) {
+  TORCH_CHECK(
+      output_edges.empty() || !accumulate_grad,
+      "outputs+accumulate_grad not yet implemented")
+  static std::mutex lock;
+  std::lock_guard<std::mutex> lock_guard(lock);
+  pybind11::gil_scoped_acquire gil;
+  NoGradGuard no_grad; // TODO(jansel): double backward
+
+  std::unordered_map<Node*, int> dependencies =
+      std::move(graph_task.dependencies_);
+  std::vector<std::shared_ptr<Node>> worklist{graph_root};
+  worklist.reserve(dependencies.size());
+  std::unordered_map<Node*, NodeCall> node_inputs;
+  node_inputs.emplace(graph_root.get(), NodeCall(graph_root));
+  std::vector<NodeCall> calls;
+  calls.reserve(dependencies.size() + 8);
+  AutogradCompilerCall compiler_call(accumulate_grad);
+  CacheNode* cache = CacheNode::root();
+
+  for (const auto i : c10::irange(output_edges.size())) {
+    auto inp = node_inputs.find(output_edges[i].function.get());
+    if (inp == node_inputs.end()) {
+      inp = node_inputs
+                .emplace(
+                    output_edges[i].function.get(),
+                    NodeCall(output_edges[i].function))
+                .first;
+    }
+    inp->second.mark_output(output_edges[i].input_nr, i);
+  }
+
+  while (!worklist.empty()) {
+    std::shared_ptr<Node> fn = std::move(worklist.back());
+    worklist.pop_back();
+
+    auto node_input_iter = node_inputs.find(fn.get());
+    TORCH_CHECK(node_input_iter != node_inputs.end());
+    NodeCall& node_call = node_input_iter->second;
+
+    { // update cache and gather args into `compiler_call`
+      CompiledNodeArgs node_args(compiler_call, node_call);
+      node_args.collect(node_call);
+      fn->compiled_args(node_args);
+      node_args.collect_hooks_from(fn.get());
+      cache = cache->lookup(node_args.key());
+    }
+
+    // finalize node construction
+    calls.emplace_back(std::move(node_call));
+    node_inputs.erase(node_input_iter);
+    size_t node_id = calls.size();
+
+    const auto& edges = fn->next_edges();
+    for (int output_id = edges.size() - 1; output_id >= 0; --output_id) {
```

Comment:
In principle the compiler can repair it, though if it is not hard to preserve the original order, it is better if we do that. Ditto with sequence IDs. We don't do *that* much reordering, for precisely the reason of memory usage.

Reply:
> though if it is not hard to preserve the original order

@jansel there's some logic here that already does graph traversal in a single threaded way that this PR can model off of
https://github.com/pytorch/pytorch/blob/e1502c0cdbfd17548c612f25d5a65b1e4b86224d/torch/csrc/autograd/engine.cpp#L400-L457

Reply:
@soulitzer I can't use that code directly because it is not symbolically keeping track of input/output relationships between nodes (which we need to generate cache keys).

I could copy and paste the `std::priority_queue` bit, though `torch.compile()` does its own reordering.  Prioritizing accumulate grads shouldn't matter because currently we are returning the needed grads from the FX graph, not doing in-place updates.

I'm not sure if it is worth taking the `O(n)` -> `O(n*log(n))` hit here to preserve an order which will prompty get ignored by `torch.compile()`.

Reply:
@soulitzer @albanD 

So I tried this as an experiment.  I believe `get_current_graph_task_execution_order()` is actually incorrectly implemented.  It works for most tests, but fails `TestCompiledAutograd::test_implicit_add`.

The issue is if you have an AccumulateGrad node with multiple incoming dependency edges.  `get_current_graph_task_execution_order()` isn't actually waiting until the dependency count reaches zero, is it relying solely on   
`sequence_nr()`.  So it ends up running AccumulateGrad after their *first* dependency is satisfied rather than their *last* one.